### PR TITLE
budget as separate contract and system call contract

### PR DESCRIPTION
**Pull request by aeyakovenko**
WIP, part of #953 
* move budget out of bank
* create a system call contract that allocates the account userdata, and assigns account's to contracts

### CODE REVIEW

**Comment by garious** [comment]
I see Instruction still depends on Budget. Does it need to?

File: src/bank.rs [hunk 1]
```diff
@@ -4,27 +4,78 @@
...
+                tokens,
+                space,
+                contract_id,
+            } => {
+                accounts[0].tokens -= tokens;
+                accounts[1].tokens += tokens;
+                if let Some(id) = contract_id {
+                    accounts[1].contract_id = id;
+                }
+                accounts[1].userdata = vec![0; space as usize];
```
Comment (mvines) [inline src/bank.rs]: should we only permit userdata if there's a `contract_id`?

File: src/bank.rs [hunk 2]
```diff
@@ -703,9 +595,11 @@ impl Bank {
         self.process_transaction(&tx).map(|_| signature)
     }
 
+    /// Each contract would need to be able to introspect its own state
+    /// this is hard coded to the budget contract langauge
     pub fn get_balance(&self, pubkey: &Pubkey) -> i64 {
         self.get_account(pubkey)
-            .map(|x| Self::get_balance_of_budget_payment_plan(&x))
+            .map(|x| budget::get_balance(&x))
```
Comment (mvines) [inline src/bank.rs]: we should check that this `Account` is a budget account first right?

File: src/bank.rs [hunk 3]
```diff
@@ -4,27 +4,78 @@
...
 use timing::{duration_as_us, timestamp};
-use transaction::{Instruction, Plan, Transaction};
+use transaction::Transaction;
 use window::WINDOW_SIZE;
 
+/// An Account with userdata that is stored on chain
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+pub struct Account {
+    /// tokens in the account
+    pub tokens: i64,
```
Comment (mvines) [inline src/bank.rs]: why do we permit negative tokens?

File: src/budget.rs [hunk 4]
```diff
@@ -110,6 +117,175 @@ impl PaymentPlan for Budget {
...
+    }
+}
+
+/// Budget DSL contract interface
+/// * tx - the transaction
+/// * accounts[0] - The source of the tokens
+/// * accounts[1] - The contract context.  Once the contract has been completed, the tokens can
+/// be spent from this account .
+pub fn process_transaction(tx: &Transaction, accounts: &mut [Account]) -> () {
+    let instruction = deserialize(&tx.userdata).unwrap();
```
Comment (mvines) [inline src/budget.rs]: worthwhile assert there are two accounts here?  I guess it'll panic in `apply_debits_to_budget_state` or `apply_credits_to_budget_state` if not...

File: src/bank.rs (hunk 3 above)
Comment (aeyakovenko) [reply src/bank.rs]: @mvines @garious had the idea that we could have validators vote with a negative fee.  so we are supporting the possibility of negative tokens.

**Comment by aeyakovenko** [comment]
@mvines @garious any ideas on naming?  budgetstate seems super lame :)

**Comment by mvines** [comment]
`BudgetState` seems fine to me for now.

File: src/bank.rs (hunk 3 above)
Comment (garious) [reply src/bank.rs]: Yeah, I still like to believe we'll be able to coalesce transactions from the right to trivialize rollback. I hijacked that top bit to keep the dream alive.

File: src/system.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
+//! system smart contract
+
+use bank::Account;
+use bincode::deserialize;
+use signature::Pubkey;
+use transaction::Transaction;
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum System {
```
Comment (jackcmay) [inline src/system.rs]: Should be called SystemContract to align with BudgetContract?

**Comment by aeyakovenko** [comment]
@mvines here is a robust transfer on date that covers trying to pay out to the wrong key
https://github.com/solana-labs/solana/pull/1189/files#diff-ea09cf31a8172005d63e915f6cc56ff7R268

File: src/bank.rs [hunk 6]
```diff
@@ -594,7 +555,7 @@ impl Bank {
         {
             let tx = &entry1.transactions[0];
             let instruction = tx.instruction();
-            let deposit = if let Instruction::NewContract(contract) = instruction {
+            let deposit = if let Some(Instruction::NewContract(contract)) = instruction {
```
Comment (mvines) [inline src/bank.rs]: Couldn't we pass `tx` through `process_transaction()` here instead of hard coding a Budget Contract instruction?

**Comment by aeyakovenko** [comment]
needs a wool update

File: src/bank.rs (hunk 6 above)
Comment (aeyakovenko) [reply src/bank.rs]: @mvines maybe, not sure.  its outside the scope of moving budget as a separate contract

**Comment by mvines** [comment]
This looks good enough to land IMO, :shipit:

**Pull request merged**